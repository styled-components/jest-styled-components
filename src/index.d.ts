interface AsymmetricMatcher {
  $$typeof: Symbol;
  sample?: string | RegExp | object | Array<any>, Function;
}

type Value = string | RegExp | AsymmetricMatcher | undefined

interface Options {
  media?: string;
  modifier?: string;
  supports?: string;
}

declare namespace jest {

  interface Matchers<R> {
    toHaveStyleRule(property: string, value: Value, options?: Options): R;
  }
}
