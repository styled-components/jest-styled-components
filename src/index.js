const { toHaveStyleRule } = require('./toHaveStyleRule');
const styleSheetSerializer = require('./styleSheetSerializer');
const { resetStyleSheet } = require('./utils');

global.beforeEach(resetStyleSheet);

expect.addSnapshotSerializer(styleSheetSerializer);
expect.extend({ toHaveStyleRule });

module.exports = {
  styleSheetSerializer,
};
