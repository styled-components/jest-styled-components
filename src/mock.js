import React from 'react';

module.exports = new Proxy(
  {},
  {
    get: function getter(target, key) {
      const StyledComponentMock = () => null;

      StyledComponentMock.displayName = `Styled${key}`;
      StyledComponentMock.extend = () => StyledComponentMock;
      StyledComponentMock.withComponent = Component => {
        const NewStyledComponent = props => <Component {...props} />;
        const suffix = Component.name ? `_${Component.name}` : '';

        NewStyledComponent.displayName = `Styled${key}${suffix}`;

        return NewStyledComponent;
      };

      return StyledComponentMock;
    },
  },
);
