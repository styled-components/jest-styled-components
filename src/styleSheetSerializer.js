const css = require('css')
const { getCSS, getHashes } = require('./utils')

const KEY = '__jest-styled-components__'

const getNodes = (node, nodes = []) => {
  if (typeof node === 'object') {
    nodes.push(node)
  }

  if (node.children) {
    Array.from(node.children).forEach(child => getNodes(child, nodes))
  }

  return nodes
}

const markNodes = nodes => nodes.forEach(node => (node[KEY] = true))

const getClassNamesFromDOM = node => Array.from(node.classList)
const getClassNamesFromProps = node => {
  const classNameProp = node.props && (node.props.class || node.props.className)

  if (classNameProp) {
    return classNameProp.trim().split(/\s+/)
  }

  return []
}

const getClassNames = nodes =>
  nodes.reduce((classNames, node) => {
    let newClassNames = null

    if (global.Element && node instanceof global.Element) {
      newClassNames = getClassNamesFromDOM(node)
    } else {
      newClassNames = getClassNamesFromProps(node)
    }

    newClassNames.forEach(className => classNames.add(className))

    return classNames
  }, new Set())

const filterClassNames = (classNames, hashes) =>
  classNames.filter(className => hashes.includes(className))

const includesClassNames = (classNames, selectors) =>
  classNames.some(className =>
    selectors.some(selector => selector.includes(className))
  )

const filterRules = classNames => rule =>
  rule.type === 'rule' &&
  includesClassNames(classNames, rule.selectors) &&
  rule.declarations.length

const getAtRules = (ast, filter) =>
  ast.stylesheet.rules
    .filter(rule => rule.type === 'media' || rule.type === 'supports')
    .reduce((acc, atRule) => {
      atRule.rules = atRule.rules.filter(filter)

      if (atRule.rules.length) {
        return acc.concat(atRule)
      }

      return acc
    }, [])

const getStyle = classNames => {
  const ast = getCSS()
  const filter = filterRules(classNames)
  const rules = ast.stylesheet.rules.filter(filter)
  const atRules = getAtRules(ast, filter)

  ast.stylesheet.rules = rules.concat(atRules)

  return css.stringify(ast)
}

const getClassNamesFromSelectorsByHashes = (classNames, hashes) => {
  const ast = getCSS()
  const filter = filterRules(classNames)
  const rules = ast.stylesheet.rules.filter(filter)

  const selectors = rules.map(rule => rule.selectors)
  const classNamesIncludingFromSelectors = new Set(classNames)
  const addHashFromSelectorListToClassNames = hash =>
    selectors.forEach(
      selectorList =>
        selectorList[0].includes(hash) &&
        classNamesIncludingFromSelectors.add(hash)
    )

  hashes.forEach(addHashFromSelectorListToClassNames)

  return [...classNamesIncludingFromSelectors]
}

const replaceClassNames = (result, classNames, style) =>
  classNames
    .filter(className => style.includes(className))
    .reduce(
      (acc, className, index) =>
        acc.replace(new RegExp(className, 'g'), `c${index++}`),
      result
    )

const replaceHashes = (result, hashes) =>
  hashes.reduce(
    (acc, className) =>
      acc.replace(
        new RegExp(`((class|className)="[^"]*?)${className}\\s?([^"]*")`, 'g'),
        '$1$3'
      ),
    result
  )

const styleSheetSerializer = {
  test(val) {
    return (
      val &&
      !val[KEY] &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (global.Element && val instanceof global.Element))
    )
  },

  print(val, print) {
    const nodes = getNodes(val)
    markNodes(nodes)

    const hashes = getHashes()
    let classNames = [...getClassNames(nodes)]
    classNames = filterClassNames(classNames, hashes)

    const style = getStyle(classNames)
    const classNamesToReplace = getClassNamesFromSelectorsByHashes(
      classNames,
      hashes
    )
    const code = print(val)

    let result = `${style}${style ? '\n\n' : ''}${code}`
    result = replaceClassNames(result, classNamesToReplace, style)
    result = replaceHashes(result, hashes)

    return result
  },
}

module.exports = styleSheetSerializer
