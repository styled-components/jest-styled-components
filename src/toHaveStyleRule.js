const { getCSS, matcherTest, buildReturnMessage } = require('./utils');

const shouldDive = node => typeof node.dive === 'function' && typeof node.type() !== 'string';

const isTagWithClassName = node => node.exists() && node.prop('className') && typeof node.type() === 'string';

const getClassNames = received => {
  let className;

  if (received) {
    if (received.$$typeof === Symbol.for('react.test.json')) {
      className = received.props.className || received.props.class;
    } else if (typeof received.exists === 'function' && received.exists()) {
      const tree = shouldDive(received) ? received.dive() : received;
      const components = tree.findWhere(isTagWithClassName);
      if (components.length) {
        className = components.first().prop('className');
      }
    } else if (global.Element && received instanceof global.Element) {
      className = Array.from(received.classList).join(' ');
    }
  }

  return className ? className.split(/\s/) : [];
};

const hasAtRule = options => Object.keys(options).some(option => ['media', 'supports'].includes(option));

const getAtRules = (ast, options) => {
  const mediaRegex = /(\([a-z-]+:)\s?([a-z0-9.]+\))/g;

  return Object.keys(options)
    .map(option =>
      ast.stylesheet.rules
        .filter(rule => rule.type === option && rule[option] === options[option].replace(mediaRegex, '$1$2'))
        .map(rule => rule.rules)
        .reduce((acc, rules) => acc.concat(rules), [])
    )
    .reduce((acc, rules) => acc.concat(rules), []);
};

const getModifiedClassName = (className, modifier = '') => {
  const classNameSelector = `.${className}`;
  let prefix = '';

  modifier = modifier.trim();
  if (modifier.includes('&')) {
    modifier = modifier.replace(/&/g, classNameSelector);
  } else {
    prefix += classNameSelector;
  }
  const first = modifier[0];
  if (first !== ':' && first !== '[') {
    prefix += ' ';
  }

  return `${prefix}${modifier}`.trim();
};

const hasClassNames = (classNames, selectors, options) =>
  classNames.some(className => selectors.includes(getModifiedClassName(className, options.modifier)));

const getRules = (ast, classNames, options) => {
  const rules = hasAtRule(options) ? getAtRules(ast, options) : ast.stylesheet.rules;

  return rules.filter(rule => rule.type === 'rule' && hasClassNames(classNames, rule.selectors, options));
};

const handleMissingRules = options => ({
  pass: false,
  message: () =>
    `No style rules found on passed Component${
      Object.keys(options).length ? ` using options:\n${JSON.stringify(options)}` : ''
    }`,
});

const getDeclaration = (rule, property) =>
  rule.declarations
    .filter(declaration => declaration.type === 'declaration' && declaration.property === property)
    .pop();

const getDeclarations = (rules, property) => rules.map(rule => getDeclaration(rule, property)).filter(Boolean);

const normalizeOptions = options =>
  options.modifier
    ? Object.assign({}, options, {
        modifier: Array.isArray(options.modifier) ? options.modifier.join('') : options.modifier,
      })
    : options;

function toHaveStyleRule(component, property, expected, options = {}) {
  const classNames = getClassNames(component);
  const ast = getCSS();
  const normalizedOptions = normalizeOptions(options);
  const rules = getRules(ast, classNames, normalizedOptions);

  if (!rules.length) {
    return handleMissingRules(normalizedOptions);
  }

  const declarations = getDeclarations(rules, property);
  const declaration = declarations.pop() || {};
  const received = declaration.value;
  const pass = !received && !expected && this.isNot ? false : matcherTest(received, expected);

  return {
    pass,
    message: buildReturnMessage(this.utils, pass, property, received, expected),
  };
}

module.exports = toHaveStyleRule;
