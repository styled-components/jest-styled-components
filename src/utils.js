const css = require('css')
const { ServerStyleSheet, isStyledComponent } = require('styled-components')

let StyleSheet

if (isStyledComponent) {
  const secretInternals = require('styled-components')
    .__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS

  if (
    secretInternals === undefined ||
    secretInternals.StyleSheet === undefined
  ) {
    throw new Error(
      'Could neither find styled-components secret internals nor styled-components/lib/models/StyleSheet.js'
    )
  } else {
    StyleSheet = secretInternals.StyleSheet
  }
} else {
  StyleSheet = require('styled-components/lib/models/StyleSheet').default // eslint-disable-line
}

const isServer = () => typeof document === 'undefined'

const resetStyleSheet = () => StyleSheet.reset(isServer())

const getHTML = () =>
  isServer()
    ? new ServerStyleSheet().getStyleTags()
    : StyleSheet.instance.toHTML()

const extract = regex => {
  let style = ''
  let matches

  while ((matches = regex.exec(getHTML())) !== null) {
    style += `${matches[1]} `
  }

  return style.trim()
}

const getStyle = () => extract(/<style[^>]*>([^<]*)</g)

const getCSS = () => css.parse(getStyle())

const getClassNames = () =>
  extract(/data-styled(?:-components)?="([^"]*)"/g).split(/\s/)

const getComponentIDs = () =>
  extract(/sc-component-id: ([^\\*\\/]*) \*\//g).split(/\s/)

const getHashes = () =>
  getClassNames()
    .concat(getComponentIDs())
    .filter(Boolean)

const buildReturnMessage = (utils, pass, property, received, expected) => () =>
  `${utils.printReceived(
    !received && !pass
      ? `Property '${property}' not found in style rules`
      : `Value mismatch for property '${property}'`
  )}\n\n` +
  'Expected\n' +
  `  ${utils.printExpected(`${property}: ${expected}`)}\n` +
  'Received:\n' +
  `  ${utils.printReceived(`${property}: ${received}`)}`

const matcherTest = (received, expected) => {
  try {
    const matcher =
      expected instanceof RegExp ? expect.stringMatching(expected) : expected

    expect(received).toEqual(matcher)
    return true
  } catch (error) {
    return false
  }
}

module.exports = {
  resetStyleSheet,
  getCSS,
  getHashes,
  buildReturnMessage,
  matcherTest,
}
