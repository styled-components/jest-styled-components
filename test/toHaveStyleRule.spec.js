import React from 'react'
import renderer from 'react-test-renderer'
import styled, { ThemeProvider, css } from 'styled-components'
import { shallow, mount } from 'enzyme'
import '../src'

const notToHaveStyleRule = (component, property, value) => {
  expect(renderer.create(component).toJSON()).not.toHaveStyleRule(
    property,
    value
  )
  expect(shallow(component)).not.toHaveStyleRule(property, value)
  expect(mount(component)).not.toHaveStyleRule(property, value)
}

const toHaveStyleRule = (component, property, value, options) => {
  expect(renderer.create(component).toJSON()).toHaveStyleRule(
    property,
    value,
    options
  )
  expect(shallow(component)).toHaveStyleRule(property, value, options)
  expect(mount(component)).toHaveStyleRule(property, value, options)
}

test('null', () => {
  expect(null).not.toHaveStyleRule('a', 'b')
})

test('non-styled', () => {
  notToHaveStyleRule(<div />, 'a', 'b')
})

test('message when rules not found', () => {
  expect(() =>
    expect(renderer.create(<div />).toJSON()).toHaveStyleRule('color', 'black')
  ).toThrowErrorMatchingSnapshot()
})

test('message when rules not found using options', () => {
  const Button = styled.button`
    color: red;
  `

  toHaveStyleRule(<Button />, 'color', 'red')
  expect(() =>
    expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
      'color',
      'red',
      {
        media: '(max-width:640px)',
        modifier: ':hover',
      }
    )
  ).toThrowErrorMatchingSnapshot()
})

test('message when property not found', () => {
  const Button = styled.button`
    color: red;
  `

  expect(() =>
    expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
      'background-color',
      'black'
    )
  ).toThrowErrorMatchingSnapshot()
})

test('message when value does not match', () => {
  const Wrapper = styled.section`
    background: orange;
  `

  expect(() => {
    expect(renderer.create(<Wrapper />).toJSON()).toHaveStyleRule(
      'background',
      'red'
    )
  }).toThrowErrorMatchingSnapshot()
})

test('basic', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `

  toHaveStyleRule(<Wrapper />, 'background', 'papayawhip')
})

test('regex', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `

  toHaveStyleRule(<Wrapper />, 'background', /^p/)
})

test('complex string', () => {
  const Wrapper = styled.section`
    border: 1px solid rgba(0, 0, 0, 0.125);
  `

  toHaveStyleRule(<Wrapper />, 'border', '1px solid rgba(0,0,0,0.125)')
})

test('undefined', () => {
  const Button = styled.button`
    cursor: ${({ disabled }) => !disabled && 'pointer'};
    opacity: ${({ disabled }) => disabled && '.65'};
  `

  toHaveStyleRule(<Button />, 'opacity', undefined)
  toHaveStyleRule(<Button />, 'cursor', 'pointer')
  toHaveStyleRule(<Button disabled />, 'opacity', '.65')
  toHaveStyleRule(<Button disabled />, 'cursor', undefined)
})

test('jest asymmetric matchers', () => {
  const Button = styled.button`
    border: 0.1em solid
      ${({ transparent }) => (transparent ? 'transparent' : 'black')};
  `

  toHaveStyleRule(<Button />, 'border', expect.any(String))
  toHaveStyleRule(<Button />, 'border', expect.stringMatching('solid'))
  toHaveStyleRule(<Button />, 'border', expect.stringMatching(/^0.1em/))
  toHaveStyleRule(<Button />, 'border', expect.stringContaining('black'))
  notToHaveStyleRule(
    <Button transparent />,
    'border',
    expect.stringContaining('black')
  )
  toHaveStyleRule(
    <Button transparent />,
    'border',
    expect.stringContaining('transparent')
  )
  notToHaveStyleRule(<Button />, 'color', expect.any(String))
  notToHaveStyleRule(<Button />, 'color', expect.anything())
})

test('any component', () => {
  const Link = ({ className, children }) => (
    <a className={className}>{children}</a>
  )

  const StyledLink = styled(Link)`
    color: palevioletred;
    font-weight: bold;
  `

  toHaveStyleRule(
    <StyledLink>Styled, exciting Link</StyledLink>,
    'color',
    'palevioletred'
  )
})

test('styled child', () => {
  const Parent = styled.div`
    color: red;
  `

  const StyledChild = styled(Parent)`
    padding: 0;
  `

  toHaveStyleRule(<StyledChild />, 'color', 'red')
})

test('extending styles', () => {
  const Button = styled.button`
    color: palevioletred;
    font-size: 1em;
    margin: 1em;
    padding: 0.25em 1em;
    border: 2px solid palevioletred;
    border-radius: 3px;
  `

  const TomatoButton = Button.extend`
    color: tomato;
    border-color: tomato;
  `

  toHaveStyleRule(<TomatoButton>Tomato Button</TomatoButton>, 'color', 'tomato')
})

test('theming', () => {
  const Button = styled.button`
    font-size: 1em;
    margin: 1em;
    padding: 0.25em 1em;
    border-radius: 3px;

    color: ${props => props.theme.main};
    border: 2px solid ${props => props.theme.main};
  `

  Button.defaultProps = {
    theme: {
      main: 'palevioletred',
    },
  }

  const theme = {
    main: 'mediumseagreen',
  }

  toHaveStyleRule(<Button>Normal</Button>, 'color', 'palevioletred')

  const component = (
    <ThemeProvider theme={theme}>
      <Button>Themed</Button>
    </ThemeProvider>
  )

  expect(renderer.create(component).toJSON()).toHaveStyleRule(
    'color',
    'mediumseagreen'
  )
  expect(mount(component)).toHaveStyleRule('color', 'mediumseagreen')
})

test('at rules', () => {
  const Wrapper = styled.section`
    color: red;
    @media (max-width: 640px) {
      color: green;
    }
    @media (min-width: 200px) and (max-width: 640px) {
      color: blue;
    }
  `

  toHaveStyleRule(<Wrapper />, 'color', 'red')
  toHaveStyleRule(<Wrapper />, 'color', 'green', {
    media: '(max-width:640px)',
  })
  toHaveStyleRule(<Wrapper />, 'color', 'green', {
    media: '(max-width: 640px)',
  })
  toHaveStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width:200px) and (max-width:640px)',
  })
  toHaveStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width: 200px) and (max-width: 640px)',
  })
  toHaveStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width: 200px) and (max-width:640px)',
  })
  toHaveStyleRule(<Wrapper />, 'color', 'blue', {
    media: '(min-width:200px) and (max-width: 640px)',
  })
})

test('selector modifiers', () => {
  const Link = styled.a`
    color: white;

    &:hover {
      color: blue;
    }
    &::after {
      color: red;
    }
    &[href*='somelink.com'] {
      color: green;
    }
    > div {
      color: yellow;
    }
    span {
      color: purple;
    }
    .child {
      color: orange;
    }
    &.self {
      color: black;
    }

    .one,
    .two {
      color: olive;
    }

    ~ div {
      &.one,
      &.two {
        color: pink;
      }
    }

    + div {
      .one,
      .two {
        color: salmon;
      }
    }

    .parent & {
      color: red;
    }

    && {
      color: fuchsia;
    }

    &&& {
      color: olive;
    }

    & & {
      color: deepskyblue;
    }
  `

  toHaveStyleRule(<Link />, 'color', 'white')
  toHaveStyleRule(<Link />, 'color', 'blue', {
    modifier: ':hover',
  })
  toHaveStyleRule(<Link />, 'color', 'red', {
    modifier: '::after',
  })
  toHaveStyleRule(<Link />, 'color', 'green', {
    modifier: "[href*='somelink.com']",
  })
  toHaveStyleRule(<Link />, 'color', 'yellow', {
    modifier: '> div',
  })
  toHaveStyleRule(<Link />, 'color', 'purple', {
    modifier: 'span',
  })
  toHaveStyleRule(<Link />, 'color', 'purple', {
    modifier: ' span',
  })
  toHaveStyleRule(<Link />, 'color', 'orange', {
    modifier: '.child',
  })
  toHaveStyleRule(<Link />, 'color', 'orange', {
    modifier: ' .child',
  })
  toHaveStyleRule(<Link />, 'color', 'black', {
    modifier: '&.self',
  })
  toHaveStyleRule(<Link />, 'color', 'olive', {
    modifier: '.one',
  })
  toHaveStyleRule(<Link />, 'color', 'olive', {
    modifier: '.two',
  })
  toHaveStyleRule(<Link />, 'color', 'pink', {
    modifier: '~ div.one',
  })
  toHaveStyleRule(<Link />, 'color', 'salmon', {
    modifier: '+ div .two',
  })
  toHaveStyleRule(<Link />, 'color', 'red', {
    modifier: '.parent &',
  })
  toHaveStyleRule(<Link />, 'color', 'fuchsia', {
    modifier: '&&',
  })
  toHaveStyleRule(<Link />, 'color', 'olive', {
    modifier: '&&&',
  })
  toHaveStyleRule(<Link />, 'color', 'deepskyblue', {
    modifier: '& &',
  })
})

test('component modifiers', () => {
  const Text = styled.span`
    color: grey;
  `

  const Link = styled.a`
    color: white;

    ${Text} {
      color: blue;
    }

    > ${Text} span {
      color: green;
    }

    ${Text} & {
      color: purple;
    }
  `

  toHaveStyleRule(<Link />, 'color', 'white')
  toHaveStyleRule(<Text />, 'color', 'grey')
  toHaveStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'blue',
    {
      // eslint-disable-next-line prettier/prettier
      modifier: css`${Text}`,
    }
  )
  toHaveStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'green',
    {
      modifier: css`> ${Text} span`,
    }
  )
  toHaveStyleRule(
    <Link>
      <Text />
    </Link>,
    'color',
    'purple',
    {
      // eslint-disable-next-line prettier/prettier
      modifier: css`${Text} &`,
    }
  )
})
