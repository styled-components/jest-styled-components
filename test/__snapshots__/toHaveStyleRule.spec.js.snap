// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message when property not found 1`] = `
"[31m\\"Property 'background-color' not found in style rules\\"[39m

Expected
  [32m\\"background-color: black\\"[39m
Received:
  [31m\\"background-color: undefined\\"[39m"
`;

exports[`message when rules not found 1`] = `"No style rules found on passed Component"`;

exports[`message when rules not found using options 1`] = `
"No style rules found on passed Component using options:
{\\"media\\":\\"(max-width:640px)\\",\\"modifier\\":\\":hover\\"}"
`;

exports[`message when value does not match 1`] = `
"[31m\\"Value mismatch for property 'background'\\"[39m

Expected
  [32m\\"background: red\\"[39m
Received:
  [31m\\"background: orange\\"[39m"
`;
